7.2.PostOffice.py

```diff
@@ -0,0 +1,86 @@
+class PostOffice:
+    """A Post Office class. Allows users to message each other.
+
+    :ivar int message_id: Incremental id of the last message sent.
+    :ivar dict boxes: Users' inboxes.
+
+    :param list usernames: Users for which we should create PO Boxes.
+    """
+
+    def __init__(self, usernames):
+        self.message_id = 0
+        self.boxes = {user: [] for user in usernames}
+
+    def send_message(self, sender, recipient, message_body, urgent=False):
+        """Send a message to a recipient.
+
+        :param str sender: The message sender's username.
+        :param str recipient: The message recipient's username.
+        :param str message_body: The body of the message.
+        :param urgent: The urgency of the message.
+        :type urgent: bool, optional
+        :return: The message ID, auto incremented number.
+        :rtype: int
+        :raises KeyError: if the recipient does not exist.
+        """
+        user_box = self.boxes[recipient]
+        self.message_id = self.message_id + 1
+        message_details = {
+            'id': self.message_id,
+            'body': message_body,
+            'sender': sender,
+            'read': False,
+        }
+        if urgent:
+            user_box.insert(0, message_details)
+        else:
+            user_box.append(message_details)
+        return self.message_id
+
+    def read_inbox(self, username, N=None):
+        """Read N first messages of username.
+
+        :param str username: Recipient's username.
+        :param N: Number of messages to read.
+        :type N: int, optional
+        :return: N unread messages.
+        :rtype: list of dict
+        :raises KeyError: if the recipient username does not exist.
+        """
+        messages = self.boxes[username]
+        if N is None:
+            N = len(messages)
+        unread = []
+        for i in range(len(messages)):
+            if N == 0:
+                break
+            if not messages[i]['read']:
+                unread.append(messages[i])
+                self.boxes[username][i]['read'] = True
+                N -= 1
+        return unread
+
+    def search_inbox(self, username, key):
+        """Search all messages of username contain key.
+
+        :param str username: Recipient's username.
+        :param str key: Keyword to search in Recipient's username messages.
+        :return: All messages contain the pass key in username .e-mail
+        :rtype: list of dict
+        :raises KeyError: if the recipient username does not exist.
+        """
+        res = []
+        messages = self.boxes[username]
+        for i in range(len(messages)):
+            if key in messages[i]['body'] or key in messages[i]['sender']:
+                res.append(messages[i])
+                self.boxes[username][i]['read'] = True
```

Comment:
1. Replace 'i' Variable names - https://workat.tech/machine-coding/tutorial/writing-meaningful-variable-names-clean-code-za4m83tiesy0 - variable names, abbreviation reference is missing

2.  You should aim to write everything in a single line, even if you can. 
This link https://towardsdatascience.com/you-should-stop-craving-for-python-one-liners-5dbf8e0de23d explains it well, even though I disagree with how important memory and runtime are in most python programs.